File: master_cic/src/master_cic_node.cpp
#include <ros/ros.h>
#include <signal.h>
#include <opencv2/opencv.hpp>
#include <image_transport/image_transport.h>
#include <opencv2/imgproc/imgproc.hpp>
#include <cv_bridge/cv_bridge.h>  
#include <sensor_msgs/image_encodings.h>
#include <ros/console.h>
#include <vector>
#include <iostream>
#include <sensor_msgs/LaserScan.h>
#include <std_msgs/String.h>
#include <std_msgs/Int16.h>
#include <std_msgs/Float32.h>

float KEEP_DIST=0.8;
bool stopFlag= false;
bool passFlag=false;
int cnt=0;
std_msgs::Int16 lineFollowSteering,lineFollowSpeed,lin_ang,lin_angs,angDir,vel,lin_dist;
std_msgs::Float32 lin_dists;
std_msgs::String lights;
class Master
{
ros::NodeHandle nh_;
image_transport::ImageTransport it_;
image_transport::Subscriber image_sub_;
image_transport::Publisher image_pub_; 
ros::Publisher pubDir,pubVel,pubLight;
ros::Subscriber steering_top,speed_top,linang_top,lindist_top,linang_tops,lindist_tops;
public:
    Master() 
    : it_(nh_)  
  {
     image_sub_ = it_.subscribe("/img_prepros", 1,&Master::imageCb, this);
     image_pub_ = it_.advertise("/master_cic", 1); 
     steering_top = nh_.subscribe("/lane_follower/steering", 1, &Master::steering_cb, this);
     speed_top= nh_.subscribe("/lane_follower/speed", 1, &Master::speed_cb, this);
     linang_top= nh_.subscribe("inter_line/ang",1,&Master::linang_cb, this);
     lindist_top=nh_.subscribe("inter_line/dist",1,&Master::lindist_cb, this);
     linang_tops= nh_.subscribe("/scan_followc/angle",1,&Master::linangs_cb, this);
     lindist_tops=nh_.subscribe("/scan_followc/dist",1,&Master::lindists_cb, this);
     pubDir= nh_.advertise<std_msgs::Int16>("/manual_control/steering",1);//Para publicar en consola
     pubVel= nh_.advertise<std_msgs::Int16>("/manual_control/speed",1);
     pubLight= nh_.advertise<std_msgs::String>("/manual_control/lights",1);
  }

int velCatch(std_msgs::Int16 vel,std_msgs::Int16 cval) 
{
 
    if(vel.data>cval.data+9)
    {
      vel.data=vel.data-2;
    }
    else if(vel.data<cval.data)
    {
      vel.data=vel.data+6;
    }
    else if(cval.data==0)
    {
      vel.data=0;
    }
    return vel.data;
}
int controlDist(std_msgs::Float32 lin_dists)
{
  int velCalc;

  if(lin_dists.data>KEEP_DIST+0.1)
  {
    velCalc= int(lin_dists.data*(-300));
  }
  else if (lin_dists.data<KEEP_DIST-0.1 && lin_dists.data>0.2)
  {
    velCalc= int(1/lin_dists.data*(50)); 
  }
  else if(KEEP_DIST-0.1<lin_dists.data && lin_dists.data<KEEP_DIST+0.1)
  {
    velCalc=0;
  }
  else
  {
    velCalc=0;
  }
  return velCalc;
}
void steering_cb(const std_msgs::Int16 steering) 
  {  
    lineFollowSteering.data = steering.data;
   // ROS_INFO("%i",lineFollowSteering.data);
  }
void speed_cb(const std_msgs::Int16 speed) 
  {  
     lineFollowSpeed.data = speed.data;
     //ROS_INFO("%i",lineFollowSpeed.data);
  }
void linang_cb(const std_msgs::Int16 linang) 
  {  
     lin_ang.data = linang.data;
  }
void lindist_cb(const std_msgs::Int16 lindist) 
  {  
     lin_dist.data=lindist.data;
   //  ROS_INFO("%i",lin_dist.data);
  }
void linangs_cb(const std_msgs::Int16 linangs) 
  { 
     lin_angs.data=linangs.data;
  }
void lindists_cb(const std_msgs::Float32 lindists) 
  {  
     lin_dists.data=lindists.data; 
  }

void imageCb(const sensor_msgs::ImageConstPtr& msg)
  {
    	cv_bridge::CvImagePtr cv_ptr;
    try
    	{
     	cv_ptr = cv_bridge::toCvCopy(msg,sensor_msgs::image_encodings::MONO8);
    	}
    catch (cv_bridge::Exception& e) 
    	{
      	ROS_ERROR("cv_bridge exception: %s", e.what());
    return;
      }
       int e1=cv::getTickCount();
///////////////////////Color transform///////////////////////////////
       cv::Mat image=cv_ptr->image;
       
     //  lights.data="diL";
       //angDir.data=98;
       //vel.data=0;
       //lineFollowSteering.data=100;
       //lineFollowSpeed.data=0;
       //lin_dists.data=0.1;
       //lin_angs.data=220;
       //lin_dist.data=100;
       //lin_ang.data=180;
/////////////////////////Overcoming//////////////////////////
       
   if (passFlag==true)
    {
       if (lin_angs.data>330 || lin_angs.data<25)
       {
            angDir.data=160;
            //pubLight.publish("diL");
            ROS_INFO("Passing left");
            lights.data="le";
       }   

       else if ( 85<lin_angs.data && lin_angs.data<220)
       {
            if (lin_dists.data>0.50)
            {
              lights.data="diL";
              passFlag=false;
              ROS_INFO("DONE passing!");
            }
            else 
            {
              lights.data="ri";
              angDir.data=37;
              ROS_INFO("Passing right");
            }
       }
       else if( 20<lin_angs.data && lin_angs.data<85)
       {
             lights.data="diL";
             angDir.data=lineFollowSteering.data;
       }
       else
       {
          ROS_INFO("B");
       }
       vel.data=-500;
    }
/////////////////////////////Lane following routine/////////////////////////////////////
    else if (stopFlag==false)
    {
      if(lin_dist.data>90)
      {
        lights.data="diL";
        angDir.data=lineFollowSteering.data;
        //gvel=lineFollowSpeed.data;
        vel.data=velCatch(vel,lineFollowSpeed);
        ROS_INFO("a: Lane Following");
      if(16>lin_angs.data || lin_angs.data>344)
        {
          angDir.data=lineFollowSteering.data;
          vel.data= controlDist(lin_dists);
          ROS_INFO("a: Obstacle Following");
          if(vel.data==0)
          {
             cnt++;
             ROS_INFO("cnt:%i",cnt);
          }
          else 
          {
               cnt=0; 
          }
          if(cnt>150)
          {  
            cnt=0;
            passFlag=true;    
          }
        }
        else 
        {  
             angDir.data=lineFollowSteering.data;
             //gvel=lineFollowSpeed.data;
             vel.data=velCatch(vel,lineFollowSpeed);
             ROS_INFO("a:Line Following");
        }
      }
      else if(15<lin_dist.data && lin_dist.data<90)
      {
          angDir.data=lineFollowSteering.data;
          vel.data=-250;
          lights.data="stop";
          ROS_INFO("a: Stopping");
          ros::Duration(0.1).sleep();
      }
      else if(15>lin_dist.data)
      {
        vel.data=0;
        angDir.data=98-lin_ang.data;
        stopFlag= true;
        lights.data="diL";
        ROS_INFO("a. Stopped");
        ros::Duration(0.1).sleep();
      }
      else
      {
        ROS_INFO("Not defined...");
      }
    }

 else if (stopFlag==true)
  {
    ROS_INFO("cnt:%i",cnt);
    if((lin_dists.data<0.2 || (30<lin_angs.data && lin_angs.data<210) || (298<lin_angs.data && lin_angs.data<330)) && cnt>5)
    {
      if (lin_dist.data<12)
      {
           stopFlag=false;
           vel.data=-300;
           cnt=0;
           lights.data="diL";
           ROS_INFO("b:Crossing DONE!");
           ros::Duration(0.8).sleep();
           //lin_dist.data=100;
           //angDir.data=98;
      }
      else
      {
         angDir.data=98-lin_ang.data;
         vel.data=-350;
         lights.data="ta";
         ROS_INFO("b:Crossing");
      }
    }
    else 
    {
        if (15>lin_dist.data)
        {
           vel.data=-100;
           pubVel.publish(vel);
           ros::Duration(0.1).sleep();
           vel.data=0;
           pubVel.publish(vel);
           lights.data="stop";
           ROS_INFO("b: Getting closer");
           
        }
        else 
        {
            vel.data=0;
            angDir.data=98;
            cnt++;
            lights.data="diL";
            ROS_INFO("b. Waiting");
         }
      ros::Duration(0.1).sleep();
    }
  }
    pubLight.publish(lights);
    pubDir.publish(angDir);
    pubVel.publish(vel);
    cv_bridge::CvImage out_msg;
    out_msg.header = msg->header;
    out_msg.encoding= sensor_msgs::image_encodings::MONO8; //BGR8
    out_msg.image= image; 
    image_pub_.publish(out_msg.toImageMsg());

int e2=cv::getTickCount();
float t=(e2-e1)/cv::getTickFrequency();
ROS_INFO("frame time: %f------------------------------------",t);        
ROS_INFO("scan_dist:%f     |       scan_ang:%d",lin_dists.data,lin_angs.data);
ROS_INFO("line_dist:%d     |       ang_Dir:%d      | vel:%d",lin_dist.data,angDir.data,vel.data); 
ROS_INFO("----------------------------------------block end");

 }
};
int main(int argc, char** argv) 
{
ros::init(argc, argv, "Master");
ROS_INFO("my_node running...");
Master ic;
//int cnt=0;
//bool stopFlag=false;
//bool passflag=false;
angDir.data=98;
vel.data=0;
lineFollowSteering.data=100;
lineFollowSpeed.data=0;
lin_dists.data=0.1;
lin_angs.data=220;
lin_dist.data=100;
lin_ang.data=180;
lights.data="dil";
ros::spin();
return 0;
}

File: master_cic/src/master_cic_node.py
#! /usr/bin/python
# it requires "chmod +x mypythonscript.py" to be called by ROS
##############################################################
# Master v2
# Autor: Cesar Bravo
# 29/03/2017
##############################################################

import rospy
from sensor_msgs.msg import CompressedImage
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError
from std_msgs.msg import Int16
from std_msgs.msg import Float32
from std_msgs.msg import String
import cv2
import numpy as np
import time
import math
from pendCalc import pend
from numpy import *

#Constants
KEEP_DIST = 0.8
bridge = CvBridge()

# Power down function
def myShut():
	
	global pubDir
	global pubAng

	pubVel.publish(0)
	pubDir.publish(98)
	time.sleep(0.5)
	print('Aufwiedersehen!')
	
# Steering callback function
def steering_callback(ros_data):

	global laneFollowSteering

	laneFollowSteering = ros_data.data

# Speed callback function
def speed_callback(ros_data):

	global laneFollowSpeed

	laneFollowSpeed = ros_data.data

# The angle of intersection line if detected detected
def line_ang_callback(ros_data):
	
	global line_ang
	
	line_ang = ros_data.data

# The distance to the intersection line  if detected detected
def line_dist_callback(ros_data):

 	global line_dist
	
	line_dist = ros_data.data

# The angle of the obstacle if detected
def scan_ang_callback(ros_data):
	
	global scan_ang
	
	scan_ang = ros_data.data

# The distance to the obstacle if detected
def scan_dist_callback(ros_data):

 	global scan_dist
	
	scan_dist = ros_data.data

# The velocity regulator function
def valCatch(vel, cval):

	if vel>cval+9:
		vel -= 2
	elif vel<cval:
		vel += 6

	elif cval == 0:
		vel = 0

	return vel

# The Following distance function
def contolDist(scan_dist):	
	
	# Approaching the obstacle (to far)
	if scan_dist > KEEP_DIST + 0.1:
		velCalc = int(scan_dist*(-300)) 
	# Range of distance to keep
	elif scan_dist < KEEP_DIST-0.1 and scan_dist>0.2:
		velCalc = int((1/scan_dist)*(50))
	# Getting away of the obstacle (to close)
	elif KEEP_DIST -0.1 < scan_dist < KEEP_DIST+0.1:
		velCalc = 0
	else: 
		velCalc = 0

	return velCalc

def image_callback(ros_data):
	
	
    try:
        # Convert your ROS Image message to OpenCV2

	# Uncomment the following line to work with raw images
        #cv2_img = bridge.imgmsg_to_cv2(ros_data, "bgr8")

	# Uncoment the following two lines to work with compressed images
	np_arr = np.fromstring(ros_data.data, np.uint8)
	cv2_img = cv2.imdecode(np_arr, cv2.CV_LOAD_IMAGE_GRAYSCALE)

    except CvBridgeError, e:
        print(e)
    else:
	
	# Processing function begins
	global pub
        global pubDir
        global pubVel	
	global angDir
	global vel	
	global laneFollowSteering
	global laneFollowSpeed
	global scan_ang
	global scan_dist
	global stopFlag
	global cnt
	global passFlag
	global lightsAnt
	global pubLight

	e1 = cv2.getTickCount()

        # 1. Convert data received to numpy array
        image = np.asarray(cv2_img) # 480x640x3	

	e1 = cv2.getTickCount()
	image = np.asarray(cv2_img) 

	lights = 'diL'
	# There exist basically three big routines within the program:
	# 1. The OVERCOMING routine---------------------------------------------------------
	if passFlag == True:
		
		# Moving to the left lane 
		if scan_ang>330 or scan_ang<25:

			angDir = 160
			pubLight.publish('diL')
			print('passing left')
			lights = 'le'

		# Retunrning to original lane
		elif 220>scan_ang>85: 

			# Ff the distance to the obstacle is greater than 
			# 0.55 m  and the angle is greater than 85 it 
              		# finishes the OVERCOMING routine
			if scan_dist>0.63:
				lights = 'diL'
				passFlag = False
				print('DONE passing!!')

			# if not, it keeps returning to the left
			else:
				lights = 'ri'
				angDir = 37
				print('passing rigth')

		# Keeps on the rigth lane until it's possible to return
		elif 85>scan_ang>20:
			lights = 'diL'
			angDir = laneFollowSteering

		else: 
			print('B')
			
		vel = -250	
		
	# 2. The LANE_FOLLOWIG routine ---------------------------------------------------------------
	elif stopFlag == False:

		# if there is no intersection line:
		if line_dist>99:
				
			lights = 'diL'
			angDir = laneFollowSteering
			gvel = laneFollowSpeed
			vel = valCatch(vel, gvel)
			print('a: Lane Following')
			
			'''if scan_ang>344 or scan_ang<16:

				angDir = laneFollowSteering
				vel = contolDist(scan_dist)
				print('a: Obstacle Following')
				
				if vel == 0:
					lights = 'stop'
					cnt += 1
					#pubVel.publish(vel)
					print(cnt)	
    					
				else:
					#pubVel.publish(vel)
					lights = 'diL'
					cnt = 0

				if cnt>60:

					cnt = 0
					passFlag = True
					
		
			else:
				lights = 'diL'
				angDir = laneFollowSteering
				gvel = laneFollowSpeed
				vel = valCatch(vel, gvel)
				#pubVel.publish(vel)
				print('a: Lane Following')'''

		elif 60>line_dist>10:	
			angDir = laneFollowSteering
			vel = -200
			#pubVel.publish(vel)
			lights = 'stop'
			print('a: Stopping...')
			time.sleep(0.1)

		elif 10>line_dist:
			vel = 0
			angDir = 98 - 2*line_ang
			#pubVel.publish(vel)
			stopFlag = True	
			lights ='diL'
			print('a: Stopped')
			time.sleep(0.3)
			
		else:
			print('Not Defined 1 : WTF??')

	elif stopFlag == True:
		
		## Existe un bug en la interseccion, cuando esta cruzando, su hay un objeto muy cerca antes de 
		## cruzar la segunda linea, no lo reconoce y se pierde
		print(cnt)
		if (scan_dist<0.2 or 210>scan_ang>30 or 298<scan_ang<330) and cnt>5:
			
			if line_dist<10:
				stopFlag = False
				vel = -300
				cnt = 0
				lights = 'dil'
				print('b: Crossing DONE!')
				time.sleep(1.0)
			else:
				angDir = 98 - 2*line_ang
				vel = -250
				lights = 'ta'
				print('b: Crossing...')
				
		else: 
			if 30>line_dist:
				vel = -100
				pubVel.publish(vel)
				lights ='stop'
				print('b: Getting closer...')
			else:
				vel = 0
				angDir = 98 
				cnt += 1
				lights ='diL'
				print('b: Watting...')

			time.sleep(0.3)
		
	
	pubDir.publish(angDir)	
	pubVel.publish(vel)

	if lights != lightsAnt:
		pubLight.publish(lights)

	lightsAnt = lights
	
	# Compress image to pub ------------------------------------------------
        cropImage = CompressedImage()
        cropImage.header.stamp = rospy.Time.now()
        cropImage.format = "jpeg"
        cropImage.data = np.array(cv2.imencode('.jpg',image)[1]).tostring()
        pub.publish(cropImage)

	# Print stats ----------------------------------------------------------
	e2 = cv2.getTickCount()	
    	t = (e2 - e1)/cv2.getTickFrequency()
	
	print('Stop: ' +str(stopFlag))
	print('scan_dist: '+str(scan_dist)+' | scan_ang: '+str(scan_ang))
	print('line_dist: '+str(line_dist)+' | angDir:'+str(angDir)+' | vel: '+str(vel)+' | frame time:'+str(t))
	print('------------------------------------------------------block end')
        

def main():

    global pub
    global pubDir
    global pubVel
    global pubLight

    global angDir
    global vel
    global cnt
    global passFlag
   
    global laneFollowSteering
    global laneFollowSpeed
    global line_ang
    global line_dist
    global stopFlag
    global scan_dist
    global scan_ang
    global lightsAnt

    angDir = 98
    vel = 0
    laneFollowSteering = 100
    laneFollowSpeed = 0
    scan_dist = 0.1
    scan_ang = 220
    line_ang = 180
    line_dist = 100
    stopFlag = False
    passFlag = False
    cnt = 0
    lightsAnt = 'diL'

    rospy.init_node('master_cic')

    image_topic = "/img_prepros/compressed"
    rospy.Subscriber(image_topic, CompressedImage, image_callback,queue_size=1)

    steering_topic = "/line_follower/steering"
    speed_topic = "/line_follower/speed"
    rospy.Subscriber(steering_topic, Int16, steering_callback, queue_size = 1)
    rospy.Subscriber(speed_topic, Int16, speed_callback, queue_size = 1)

    line_ang_topic = "/inter_line/ang"
    line_dist_topic = "/inter_line/dist"  
    rospy.Subscriber(line_ang_topic, Int16, line_ang_callback, queue_size = 1)
    rospy.Subscriber(line_dist_topic, Int16, line_dist_callback, queue_size = 1)

    line_ang_topic = "/scan_follow/angle"
    line_dist_topic = "/scan_follow/dist"  
    rospy.Subscriber(line_ang_topic, Int16, scan_ang_callback, queue_size = 1)
    rospy.Subscriber(line_dist_topic, Float32, scan_dist_callback,queue_size=1)

    pub = rospy.Publisher('/master/compressed',CompressedImage, queue_size=1)
    pubDir = rospy.Publisher('/manual_control/steering', Int16, queue_size=1)
    pubVel = rospy.Publisher('/manual_control/speed', Int16, queue_size=1)
    pubLight = rospy.Publisher('/manual_control/lights', String, queue_size=1)

    rospy.spin()
    rospy.on_shutdown(myShut)
    

if __name__ == '__main__':
    main()

